packages/agent/src/oidc-v2.ts

```diff
@@ -0,0 +1,689 @@
+   * for processing. */
+  exp: number;
+  /** Time when the End-User authentication occurred. */
+  auth_time?: number;
+  /** b64url encoded nonce used to associate a Client session with an ID Token, and to
+   * mitigate replay attacks. */
+  nonce?: string;
+  /** Custom claims. */
+  [key: string]: any;
+};
+
+/** The fields for an Web5 Connect Auth Response */
+export type Web5ConnectAuthResponse = {
+  delegation_grants: any[];
+  privateKeyJwks: Jwk[];
+};
+
+/** Represents the different OIDC endpoint types.
+ * 1. `pushedAuthorizationRequest`: client sends {@link PushedAuthRequest} receives {@link PushedAuthResponse}
+ * 2. `authorize`: provider gets the {@link HybridAuthRequest} JWT that was stored by the PAR
+ * 3. `callback`: provider sends {@link HybridAuthResponse} to this endpoint
+ * 4. `token`: client gets {@link HybridAuthResponse} from this endpoint
+ */
+type OidcEndpoint =
+  | 'pushedAuthorizationRequest'
+  | 'authorize'
+  | 'callback'
+  | 'token';
+
+/**
+ * Gets the correct OIDC endpoint out of the {@link OidcEndpoint} options provided.
+ * Handles a trailing slash on baseURL
+ *
+ * @param {Object} options the options object
+ * @param {string} options.baseURL for example `http://foo.com/connect/
+ * @param {OidcEndpoint} options.endpoint the OIDC endpoint desired
+ * @param {string} options.authParam this is the unique id which must be provided when getting the `authorize` endpoint
+ * @param {string} options.tokenParam this is the random state as b64url which must be provided with the `token` endpoint
+ */
+function buildOidcUrl({
+  baseURL,
+  endpoint,
+  authParam,
+  tokenParam,
+}: {
+  baseURL: string;
+  endpoint: OidcEndpoint;
+  authParam?: string;
+  tokenParam?: string;
+}) {
+  switch (endpoint) {
+    /** 1. client sends {@link PushedAuthRequest} & receives {@link PushedAuthResponse} */
+    case 'pushedAuthorizationRequest':
+      return appendPathToUrl({
+        path : 'par',
+        url  : baseURL,
+      });
+    /** 2. provider gets {@link HybridAuthRequest} */
+    case 'authorize':
+      return authParam
+        ? appendPathToUrl({
+          path : `authorize/${authParam}.jwt`,
+          url  : baseURL,
+        })
+        : '';
+    /** 3. provider sends {@link HybridAuthResponse} */
+    case 'callback':
+      return appendPathToUrl({
+        path : `callback`,
+        url  : baseURL,
+      });
+    /**  4. client gets {@link HybridAuthResponse */
+    case 'token':
+      return tokenParam
+        ? appendPathToUrl({
+          path : `token/${tokenParam}.jwt`,
+          url  : baseURL,
+        })
+        : '';
+    // TODO: metadata endpoints?
+    default:
+      return '';
+  }
+}
+
+/**
+ * Generates a random string value used to associate a Client authorization
+ * request with the Identity Provider's response callback.
+ *
+ * @returns A random state as a Base64Url encoded string.
+ */
+function generateRandomState() {
```

Comment:
Consider generalizing for reuse and readability:

```suggestion
function generateRandomBase64Url(byteLength: number):  { ... }
```

Reply:
No need for this, it's the way it is because its locked to 12 bytes

Reply:
For my learning:

1. Is state required to be random bytes? This is different to what I normally understood to be the intent of `state` in OAuth/OIDC.
1. Is the 12 bytes a spec compliance thing, or something that has enough entropy?
1. Is the `nonce` required to be derived `state`?

Reply:
Good catch actually. I misread the spec and it was poorly written:

> The nonce parameter value needs to include per-session state and be unguessable to attackers. One method to achieve this for Web Server Clients is to store a cryptographically random value as an HttpOnly session cookie and use a cryptographic hash of the value as the nonce parameter. In that case, the nonce in the returned ID Token is compared to the hash of the session cookie to detect ID Token replay by third parties. A related method applicable to JavaScript Clients and other Browser-Based Clients is to store the cryptographically random value in HTML5 local storage and use a cryptographic hash of this value.

By "include per-session state" they just mean the nonce should be a nonce; which is tautological and unnecessary to write. I solved for your feedback and removed all of the abstraction here.

Reply:
So for my understanding: is the 12 bytes a spec compliance thing?

If it's not a spec thing, then seems odd to choose 12. Conventionally 16 or 32 bytes (128 or 256 bits) are used.

Reply:
fixed